### PR TITLE
Recommend `via-ir` whenever a stack too deep error is encountered.

### LINKED ISSUE #12912: Suggest trying to compile via IR in the stack-too-deep error message in legacy compilation.
Since it seems like via IR compilation works in a lot of such cases these days, this might help against stack-too-deep frustration and help in promoting via-IR codegen.

### PR DESCRIPTION
**Pull request by hrkrshnn**
Closes https://github.com/ethereum/solidity/issues/12912

### CODE REVIEW

File: libsolutil/StackTooDeepString.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#pragma once
+
+#include <string>
+
+namespace solidity::util
+{
+	static std::string stackTooDeepString =
+		"Stack too deep. "
+		"Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON). "
+		"Otherwise, try removing local variables.";
```
Comment (leonardoalt) [inline libsolutil/StackTooDeepString.h]: This removes the information of how much deeper than allowed it actually is.
I'm fine with that, but consensus?

**Comment by leonardoalt** [comment]
I guess it needs a rebase for the smt tests?

Also, do we not have tests for this?!

File: libsolidity/codegen/CompilerContext.cpp [hunk 2]
```diff
@@ -422,7 +423,7 @@ void CompilerContext::appendInlineAssembly(
 			BOOST_THROW_EXCEPTION(
 				StackTooDeepError() <<
 				errinfo_sourceLocation(nativeLocationOf(_identifier)) <<
-				util::errinfo_comment("Stack too deep (" + to_string(stackDiff) + "), try removing local variables.")
```
Comment (chriseth) [inline libsolidity/codegen/CompilerContext.cpp:425]: If this is always using the same text, does it make sense to add the comment in the constructor of the error?

File: libsolutil/StackTooDeepString.h (hunk 1 above)
Comment (chriseth) [reply libsolutil/StackTooDeepString.h]: I also think it's not such a useful information.
But I think we should suggest to also activate the optimizer, shouldn't we?

Comment (hrkrshnn) [reply libsolutil/StackTooDeepString.h]: Just added info about optimizer.

File: libsolidity/codegen/CompilerContext.cpp (hunk 2 above)
Comment (hrkrshnn) [reply libsolidity/codegen/CompilerContext.cpp:425]: What do you mean?

**Review by ekpyron — APPROVED** [review]
Something like https://github.com/ethereum/solidity/pull/12983/files#r865705579 would be nice, but at a quick glance, I'm not sure if it'd still properly play together e.g. with ``assertThrow`` and such...
So I'd say let's just merge this as is - might be nice to have in the upcoming release.

**Comment by ekpyron** [comment]
> Also, do we not have tests for this?!

Technically, it's an ICE and we don't test for ICEs, since we usually fix them instead :-D...
Although we might want to think about dragging a proper error reporter to codegen for this... then we could also start emitting info messages about stack limit evasion and such...

**Comment by ekpyron** [comment]
Damn, I thought I merged this before the release... merging it now anyways - we can still do more cosmetics afterwards, but we should definitely have this suggestion in general...

**Pull request merged**